File: packages/common/src/context.ts
import { AsyncLocalStorage } from 'async_hooks';
import { v4 as uuidv4 } from 'uuid';

/**
 * Context data stored in AsyncLocalStorage
 */
export interface RequestContext {
  requestId: string;
  orgId?: string;
  tenantId?: string;
  userId?: string;
  timestamp: Date;
  metadata?: Record<string, any>;
}

/**
 * AsyncLocalStorage for request context
 */
const contextStorage = new AsyncLocalStorage<RequestContext>();

/**
 * Initialize request context with a new request ID
 */
export function initializeContext(
  requestId?: string,
  data?: Partial<Omit<RequestContext, 'requestId' | 'timestamp'>>
): RequestContext {
  const context: RequestContext = {
    requestId: requestId || uuidv4(),
    timestamp: new Date(),
    ...data
  };
  return context;
}

/**
 * Run function within a new request context
 */
export function runWithContext<T>(
  context: RequestContext,
  fn: () => T
): T {
  return contextStorage.run(context, fn);
}

/**
 * Get current request context
 */
export function getContext(): RequestContext | undefined {
  return contextStorage.getStore();
}

/**
 * Get current request ID
 */
export function getRequestId(): string | undefined {
  const context = getContext();
  return context?.requestId;
}

/**
 * Get current organization ID
 */
export function getOrgId(): string | undefined {
  const context = getContext();
  return context?.orgId;
}

/**
 * Get current tenant ID
 */
export function getTenantId(): string | undefined {
  const context = getContext();
  return context?.tenantId;
}

/**
 * Get current user ID
 */
export function getUserId(): string | undefined {
  const context = getContext();
  return context?.userId;
}

/**
 * Update current context with partial data
 */
export function updateContext(data: Partial<RequestContext>): void {
  const context = getContext();
  if (context) {
    Object.assign(context, data);
  }
}

/**
 * Express/Connect middleware to initialize request context
 */
export function contextMiddleware() {
  return (req: any, res: any, next: any) => {
    const requestId = req.headers['x-request-id'] as string || uuidv4();
    const orgId = req.headers['x-org-id'] as string;
    const tenantId = req.headers['x-tenant-id'] as string;
    
    const context = initializeContext(requestId, {
      orgId,
      tenantId,
      userId: req.user?.id
    });

    // Set response header
    res.setHeader('x-request-id', context.requestId);

    runWithContext(context, () => {
      next();
    });
  };
}

File: packages/common-py/src/finops_common/context.py
"""Request context management using context variables."""
import uuid
from contextvars import ContextVar
from typing import Optional

# Context variables for request tracking
_request_id_var: ContextVar[Optional[str]] = ContextVar("request_id", default=None)
_org_id_var: ContextVar[Optional[str]] = ContextVar("org_id", default=None)
_tenant_id_var: ContextVar[Optional[str]] = ContextVar("tenant_id", default=None)
_user_id_var: ContextVar[Optional[str]] = ContextVar("user_id", default=None)


def set_request_id(request_id: Optional[str] = None) -> str:
    """Set request ID in context. Generate new UUID if not provided."""
    if request_id is None:
        request_id = str(uuid.uuid4())
    _request_id_var.set(request_id)
    return request_id


def get_request_id() -> Optional[str]:
    """Get current request ID from context."""
    return _request_id_var.get()


def set_org_id(org_id: Optional[str]) -> None:
    """Set organization ID in context."""
    _org_id_var.set(org_id)


def get_org_id() -> Optional[str]:
    """Get current organization ID from context."""
    return _org_id_var.get()


def set_tenant_id(tenant_id: Optional[str]) -> None:
    """Set tenant ID in context."""
    _tenant_id_var.set(tenant_id)


def get_tenant_id() -> Optional[str]:
    """Get current tenant ID from context."""
    return _tenant_id_var.get()


def set_user_id(user_id: Optional[str]) -> None:
    """Set user ID in context."""
    _user_id_var.set(user_id)


def get_user_id() -> Optional[str]:
    """Get current user ID from context."""
    return _user_id_var.get()


def clear_context() -> None:
    """Clear all context variables."""
    _request_id_var.set(None)
    _org_id_var.set(None)
    _tenant_id_var.set(None)
    _user_id_var.set(None)
